File: Appstax/Appstax/AXLoginUIManager.m

#import "AppstaxInternals.h"
#import "AXLoginUIManager.h"
#import "AXLoginViewController.h"

@interface AXLoginUIManager ()
@property AXLoginViewController *signupViewController;
@property AXLoginViewController *loginViewController;
@property (weak) AXUserService *userService;
@property (weak) UIViewController *presentationRoot;
@property UINavigationController *navigationController;
@property (copy) void (^completion)(void);
@end

@implementation AXLoginUIManager

- (instancetype)initWithUserService:(AXUserService *)userService {
    self = [super init];
    if(self) {
        _userService = userService;
        [self setupSignupViewController];
        [self setupLoginViewController];
        [self setupNavigationController];
    }
    return self;
}

- (void)setupNavigationController {
    _navigationController = [[UINavigationController alloc] initWithRootViewController:_signupViewController];
    _navigationController.navigationBarHidden = YES;
}

- (void)setupSignupViewController {
    _signupViewController = [[AXLoginViewController alloc] initWithUIManager:self];
    _signupViewController.submitTitle = @"Sign up";
}

- (void)setupLoginViewController {
    _loginViewController = [[AXLoginViewController alloc] initWithUIManager:self];
    _loginViewController.submitTitle = @"Log in";
    _loginViewController.goToLoginHidden = YES;
}

- (void)setupCustomViews:(void(^)(AXLoginViews *views))loginViews {
    AXLoginViews *views = [[AXLoginViews alloc] initWithSize:_loginViewController.view.bounds.size];
    loginViews(views);
    [_signupViewController setBackgroundView:views.signup];
    [_loginViewController setBackgroundView:views.login];
}

- (void)presentModalLoginWithViews:(void(^)(AXLoginViews *views))loginViews completion:(void(^)(void))completion {
    self.completion = completion;
    dispatch_async(dispatch_get_main_queue(), ^{
        _presentationRoot = [UIApplication sharedApplication].keyWindow.rootViewController;
        if(loginViews) {
            [self setupCustomViews:loginViews];
        }
        [_presentationRoot presentViewController:_navigationController animated:YES completion:nil];
    });
}

- (void)viewControllerDidPressSubmitButton:(UIViewController *)viewController {
    if([viewController isEqual:_signupViewController]) {
        [self signupViewControllerPressedSubmitButton];
    }
    if([viewController isEqual:_loginViewController]) {
        [self loginViewControllerPressedSubmitButton];
    }
}

- (void)viewControllerDidPressGoToLoginButton:(UIViewController *)viewController {
    [_navigationController pushViewController:_loginViewController animated:YES];
}

- (void)signupViewControllerPressedSubmitButton {
    [_userService signupWithUsername:_signupViewController.username
                            password:_signupViewController.password
                          completion:^(AXUser *user, NSError *error) {
                              if(!error) {
                                  [self finish];
                              } else {
                                  [_signupViewController showError:error.userInfo[@"ErrorMessage"]];
                              }
                          }];
}

- (void)loginViewControllerPressedSubmitButton {
    [_userService loginWithUsername:_loginViewController.username
                           password:_loginViewController.password
                         completion:^(AXUser *user, NSError *error) {
                             if(!error) {
                                 [self finish];
                             } else {
                                 [_loginViewController showError:error.userInfo[@"ErrorMessage"]];
                             }
                         }];
}

- (void)finish {
    [_presentationRoot dismissViewControllerAnimated:YES completion:^{
        [_loginViewController clear];
        [_signupViewController clear];
    }];
    _completion();
}

@end

File: Appstax/Appstax/AXLoginUIManager.swift

import UIKit

class AXLoginUIManager {
    
    private var signupViewController: AXLoginViewController!
    private var loginViewController: AXLoginViewController!
    private var userService: AXUserService
    private var presentationRoot: UIViewController!
    private var navigationController: UINavigationController!
    private var completion: (()->())?
    
    init(userService: AXUserService) {
        self.userService = userService
        
        signupViewController = AXLoginViewController(manager: self)
        signupViewController.submitTitle = "Sign up"
        
        loginViewController = AXLoginViewController(manager: self)
        loginViewController.submitTitle = "Log in"
        loginViewController.goToLoginHidden = true
        
        navigationController = UINavigationController(rootViewController: signupViewController)
        navigationController.navigationBarHidden = true
    }
    
    func runConfig(run:((AXLoginConfig)->())?) {
        let views = AXLoginViews(size: loginViewController.view.bounds.size)
        let config = AXLoginConfig(views: views, providers: [])
        run?(config)
        signupViewController.backgroundView = views.signup
        signupViewController.providers = config.providers
        loginViewController.backgroundView = views.login
    }
    
    func presentModalLoginWithConfig(config:((AXLoginConfig)->())?, completion:()->()) {
        self.completion = completion
        dispatch_async(dispatch_get_main_queue()) {
            self.presentationRoot = UIApplication.sharedApplication().keyWindow?.rootViewController
            self.runConfig(config)
            self.presentationRoot.presentViewController(self.navigationController, animated: true, completion: nil)
        }
    }
    
    func viewControllerDidPressSubmitButton(viewController: AXLoginViewController) {
        if viewController === signupViewController {
            signupViewControllerDidPressSubmitButton()
        }
        if viewController === loginViewController {
            loginViewControllerDidPressSubmitButton()
        }
    }
    
    func viewControllerDidPressProviderButton(viewController: AXLoginViewController, provider: String) {
        userService.login(provider: provider, fromViewController: viewController) {
            user, error in
            if let error = error {
                viewController.showError(error.userInfo["errorMessage"] as? String ?? "")
            } else {
                self.finish()
            }
        }
    }
    
    func viewControllerDidPressGoToLoginButton(viewController: AXLoginViewController) {
        navigationController.pushViewController(loginViewController, animated: true)
    }
    
    func signupViewControllerDidPressSubmitButton() {
        userService.signup(username: signupViewController.username, password: signupViewController.password, login: true, properties: [:]) {
            user, error in
            if let error = error {
                self.signupViewController.showError(error.userInfo["errorMessage"] as? String ?? "")
            } else {
                self.finish()
            }
        }
    }
    
    func loginViewControllerDidPressSubmitButton() {
        userService.login(username: loginViewController.username, password: loginViewController.password) {
            user, error in
            if let error = error {
                self.loginViewController.showError(error.userInfo["errorMessage"] as? String ?? "")
            } else {
                self.finish()
            }
        }
    }
    
    func finish() {
        presentationRoot.dismissViewControllerAnimated(true) {
            self.loginViewController.clear()
            self.signupViewController.clear()
        }
        completion?()
    }
}